src/compiler/propagation/type_propagator.cc

```diff
@@ -388,9 +388,45 @@ void TypePropagator::call_method(MethodTemplate* caller,
 
 void TypePropagator::call_static(MethodTemplate* caller, TypeScope* scope, uint8* site, Method target) {
   TypeStack* stack = scope->top();
+  int arity = target.arity();
+
   std::vector<ConcreteType> arguments;
   stack->push_empty();
-  call_method(caller, scope, site, target, arguments);
+
+  bool is_static = true;
+  int offset = target.selector_offset();
+  if (offset >= 0) {
+    TypeSet receiver = stack->local(arity);
+    // If the receiver is a single type, we don't need
+    // to do virtual lookups. This allows us to handle
+    // super calls where we cannot find the (virtual)
+    // super method in the dispatch table entries for
+    // the receiver because they have been overridden.
+    is_static = receiver.size(words_per_type_) <= 1;
+  }
+
+  if (is_static) {
+    call_method(caller, scope, site, target, arguments);
+  } else {
+    Program* program = this->program();
```

Comment:
Wouldn't it be easier to just call `call_virtual` from here?
Actually, from line 407, and returning, so that you can move the `stack->push_empty from line 394 to after the `call_virtual(...); return`.

If we are doing something different than `call_virtual` add a comment explaining why that's not what we are doing.